ephios/core/views/event.py

```diff
@@ -609,30 +602,43 @@ def form_valid(self, form):
         messages.success(self.request, _("Event copied successfully."))
         return redirect(reverse("core:event_list"))
 
+    @classmethod
+    def as_view(cls, **initkwargs):
+        return csp_exempt(super().as_view(**initkwargs))
 
-class RRuleOccurrenceView(CustomPermissionRequiredMixin, View):
-    permission_required = "core.add_event"
 
-    def post(self, *args, **kwargs):
-        try:
-            recurrence = RecurrenceField().clean(self.request.POST["recurrence_string"])
-            return HttpResponse(
-                json.dumps(
-                    recurrence.between(
-                        datetime.now() - timedelta(days=1),
-                        datetime.now()
-                        + timedelta(days=7305),  # allow dates up to twenty years in the future
-                        inc=True,
-                        dtstart=datetime.combine(
-                            forms.DateField().to_python(self.request.POST["dtstart"]),
-                            datetime.min.time(),
-                        ),
-                    ),
-                    default=lambda obj: date_format(obj, format="SHORT_DATE_FORMAT"),
-                )
-            )
-        except (TypeError, KeyError, ValidationError):
-            return HttpResponse()
+class ShiftCopyView(CustomPermissionRequiredMixin, SingleObjectMixin, FormView):
+    permission_required = "core.change_event"
+    model = Shift
+    template_name = "core/shift_copy.html"
+    form_class = ShiftCopyForm
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.object = self.get_object()
+
+    def get_success_url(self):
+        return reverse("core:event_detail", kwargs={"pk": self.object.event.pk, "slug": "event"})
+
+    def form_valid(self, form):
+        shift = self.object
+        duration = shift.end_time - shift.start_time
+        meeting_offset = shift.start_time - shift.meeting_time
+        shifts_to_create = []
+        for dt in form.cleaned_data["recurrence"].xafter(timezone.now(), 1000, inc=True):
+            shift = copy(shift)
+            shift.pk = None
+            shift.meeting_time = dt - meeting_offset
+            shift.start_time = dt
+            shift.end_time = dt + duration
```

Comment:
I get the wrong times when copying shifts (UTC instead of correct timezone?!)

Reply:
My system time is Europe/Berlin. I have no timezone issues on other sites. 🤔
If you can't reproduce it locally, I can look into it when I'm back at that computer.

Reply:
hm fair point. it was easier this way as handling dates in timezones different from the browser in JS is a huge pain. as it was working for Lukas as well as for me it would be cool if you could look into it a bit on your machine to save us the major rewrite ^^

Reply:
So it is showing the correct times in the preview:
![image](https://github.com/user-attachments/assets/8e4d1b50-040d-40ca-b4a8-7b71595b536c)
but it saves the wrong time
![image](https://github.com/user-attachments/assets/72d69433-a985-4f76-8789-daf8d4cb8db9)

Reply:
![image](https://github.com/user-attachments/assets/670fbc34-1b70-46f1-981b-0599c52a45e3)
and it seems my browser is timezone aware

Reply:
I debugged through it and I think that in ShiftCopyView you are interpreting the localized times from the form input as UTC. It shoud work if you just assign the right timezone :thinking:

Reply:
does it work now?

Reply:
Nope.

Reply:
Write a test to prove it 😅

Reply:
Now I'm facing the opposite problem: When using single dates, the times in the preview are wrong. The actual copy operation creates the copies at the correct time.

![Screenshot 2024-12-13 at 12-14-12 ephios - Copy shift](https://github.com/user-attachments/assets/763213f5-8c96-4e7f-9f36-d2839aa9533f)

What surprises me is that there is an offset of +2, while my browser timezone is set to UTC+1

Reply:
For reference: This does also happen when using hourly/daily/weekly/monthly/yearly rules.

> What surprises me is that there is an offset of +2, while my browser timezone is set to UTC+1

My bad. I was using an event from your test data on June 30, so an offset of +2 makes sense (daylight saving time).